src/server.c

```diff
@@ -358,11 +350,23 @@ static bool handle_server_work(PgSocket *server, PktHdr *pkt)
 		}
 		/* ErrorResponse and CommandComplete show end of copy mode */
 		if (server->copy_mode) {
+			slog_debug(server, "COPY failed");
 			server->copy_mode = false;
 
-			/* it's impossible to track sync count over copy */
-			if (client)
-				client->expect_rfq_count = 0;
+			/*
+			 * Clear until next CopyDone or CopyFail message in the
+			 * queue. This is needed to remove any Sync messages
+			 * from the outstanding requests queue, for which we
+			 * don't expect a response from the server.
+			 *
+			 * It isn't a problem if the CopyDone or CopyFail
+			 * message has not been received yet. This message will
+			 * be removed from the queue later when the server
+			 * sends a ReadyForQuery message and we clear the queue
+			 * until the next Sync.
+			 */
+			if (!clear_outstanding_requests_until(server, "cf"))
+				return false;
 		} else {
 			server->query_failed = true;
```

Comment:
This should happen always probably, otherwise we won't pop a later received CopyDone message in case of failure.